include/RAJA/pattern/teams/teams_openmp.hpp

```diff
@@ -104,6 +104,76 @@ struct LoopExecute<omp_parallel_for_exec, SEGMENT> {
   }
 };
 
+  //
+  // Return local index
+  //
+template <typename SEGMENT>
+struct LoopICountExecute<omp_parallel_for_exec, SEGMENT> {
+
+  template <typename BODY>
+  static RAJA_INLINE RAJA_HOST_DEVICE void exec(
+      LaunchContext const RAJA_UNUSED_ARG(&ctx),
+      SEGMENT const &segment,
+      BODY const &body)
+  {
+
+    int len = segment.end() - segment.begin();
+#pragma omp parallel for
+    for (int i = 0; i < len; i++) {
```

Comment:
Would it be better to run the loop from begin() to end() and avoid the offset calculation in the loop, then use a local counter variable for the second argument to body(); i.e., int ii = 0 outside of loop, then do ii++ in the loop? I am wondering whether the compiler would optimize better....

Same question in other places in the code.

Reply:
@rhornung67 , I think that would end up in a race condition for ii with omp. One option could be to make change it to 

```
    RAJA_EXTRACT_BED_IT(segment);                                                                                                                              
                                                                                                                                                               
#pragma omp parallel for                                                                                                                                       
    for (decltype(distance_it) i = 0; i < distance_it; ++i) {                                                                                                  
                                                                                                                                                               
      body(*(begin_it + i), i);                                                                                                                                
    } 
```
    
Which is more consistent with the rest of RAJA, would this work?

Reply:
That's a good point about a race condition since the declaration and initialization would occur inside an openmp parallel region. We should try a few things with different compilers to see how they impact performance. I was thinking in terms of trying to make the code as simple as possible for compilers to optimize.

Reply:
When I ran some of these numbers before, some of the compilers strongly preferred having an integer rather than iterator as their loop variable.  Others it didn't matter, but the version there worked well in general.  It wouldn't hurt to see what happens with another round of tests though, if we could get rid of that extra boiler-plate and directly use the iterators I'd be pretty happy about it, especially if the compilers did well with the range-based for loop (which wasn't supported in OpenMP yet when we did that refactor so it wasn't portable):

```c++
for(auto & i : segment) {
  body(i);
}
```

Reply:
Yes, I wasn't clear in my original comment. I meant to do something like 

int ibeg = segment.begin();
int iend = segment.end();
for (int i = ibeg; i < iend; ++i) ...

Reply:
That's what the BED_IT one does though, right? To get from an iterator to what we need to give the body requires either an offset and dereference or `begin_it[I]`, which should usually be (but is not always) the same.